test/propagated_params_test.go

```diff
@@ -0,0 +1,382 @@
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	runtime "k8s.io/apimachinery/pkg/runtime"
+	"knative.dev/pkg/apis"
+	duckv1beta1 "knative.dev/pkg/apis/duck/v1beta1"
+	knativetest "knative.dev/pkg/test"
+)
+
+var (
+	ignoreTypeMeta          = cmpopts.IgnoreFields(metav1.TypeMeta{}, "Kind", "APIVersion")
+	ignoreObjectMeta        = cmpopts.IgnoreFields(metav1.ObjectMeta{}, "ResourceVersion", "UID", "CreationTimestamp", "Generation", "ManagedFields", "Labels")
+	ignoreCondition         = cmpopts.IgnoreFields(apis.Condition{}, "LastTransitionTime.Inner.Time", "Message")
+	ignorePipelineRunStatus = cmpopts.IgnoreFields(v1beta1.PipelineRunStatusFields{}, "StartTime", "CompletionTime", "FinallyStartTime")
+	ignoreTaskRunStatus     = cmpopts.IgnoreFields(v1beta1.TaskRunStatusFields{}, "StartTime", "CompletionTime")
+	ignoreConditions        = cmpopts.IgnoreFields(duckv1beta1.Status{}, "Conditions")
+	ignoreContainerStates   = cmpopts.IgnoreFields(corev1.ContainerState{}, "Terminated")
+	ignoreStepState         = cmpopts.IgnoreFields(v1beta1.StepState{}, "ImageID")
+)
+
+func TestPropagatedParams(t *testing.T) {
+	t.Parallel()
+	type tests struct {
+		name            string
+		pipelineName    string
+		taskName        string
+		pipelineRunFunc func(*testing.T, string) (*v1beta1.PipelineRun, *v1beta1.PipelineRun)
+	}
+
+	tds := []tests{{
+		name:            "propagated parameters fully",
+		pipelineName:    "propagated-parameters-fully",
+		taskName:        "echo-hello",
+		pipelineRunFunc: getPropagatedParamPipelineRun,
+	}, {
+		name:            "propagated parameters with inner scope",
+		pipelineName:    "propagated-parameters-inner-scope",
+		taskName:        "echo-hello",
+		pipelineRunFunc: getPropagatedParamWithScopePipelineRun,
+	}, {
+		name:            "propagated parameters with default inner scope",
+		pipelineName:    "propagated-parameters-default-inner-scope",
+		taskName:        "echo-hello",
+		pipelineRunFunc: getPropagatedParamWithDefaultScopePipelineRun,
+	}}
+
+	for _, td := range tds {
+		td := td
+		t.Run(td.name, func(t *testing.T) {
+			t.Parallel()
+			ctx := context.Background()
+			ctx, cancel := context.WithCancel(ctx)
+			defer cancel()
+
+			propagatedParamFlags := requireAllGates(map[string]string{
+				"enable-api-fields": "alpha",
+			})
+			c, namespace := setup(ctx, t, propagatedParamFlags)
+
+			embeddedStatus := GetEmbeddedStatus(ctx, t, c.KubeClient)
+			knativetest.CleanupOnInterrupt(func() { tearDown(ctx, t, c, namespace) }, t.Logf)
+			defer tearDown(ctx, t, c, namespace)
+
+			t.Logf("Setting up test resources for %q test in namespace %s", td.name, namespace)
+			pipelineRun, expectedResolvedPipelineRun := td.pipelineRunFunc(t, namespace)
+			if embeddedStatus == config.MinimalEmbeddedStatus {
+				expectedResolvedPipelineRun.Status.ChildReferences = []v1beta1.ChildStatusReference{}
+				for k, v := range expectedResolvedPipelineRun.Status.TaskRuns {
+					k := k
```

Comment:
I don't think this is necessary since this is running in the same goroutine unlike the above for loop where each t.Run is running in parallel

Reply:
Hmm. I did try it without that and ran into errors occasionally when I ran it multiple times. I thought this might have had something to do with it. I can retry a few more times to validate that.

Reply:
You are right. I get this error even after setting `k := k, ...` because the order of the taskruns (the the case where I have a regular and a finally task changes) and so when I compare, there is a difference. I will fix it. Thanks!

Reply:
fixed this.